File: Problem6/haskell/sum_square_difference.hs
sum_square_difference :: Integer -> Integer
sum_square_difference x = square_of_the_sum x - sum_of_the_squares[1..x]

sum_of_the_squares :: [Integer] -> Integer
sum_of_the_squares [] = 0
sum_of_the_squares (x:xs) = (x * x) + sum_of_the_squares xs 

square_of_the_sum :: Integer -> Integer
square_of_the_sum x = sum[1..x] * sum[1..x]

File: Problem6/ruby/lib/sum_square_difference.rb
class SumSquareDifference
  def sum_square_difference x
    range = (1..x).to_a
    square_of_the_sum(range) - sum_of_the_squares(range)
  end
  
  def sum_of_the_squares x
    squares = x.map do |n|
      n * n
    end
    squares.reduce(:+)
  end

  def square_of_the_sum x
    sum = x.reduce(:+)
    sum * sum
  end
end
